### DISCUSSION TITLE
Token decoding without a route

### DISCUSSION THREAD

**Discussion by asta-nguyen** [adonisjs/core#3597] [Help]
I'm attempting to decode a token in order to obtain user information. However, after reading the documentation, I discovered that we can obtain user information using route authentication as 
```js
import Route from '@ioc:Adonis/Core/Route'
Route.get('dashboard', async ({ auth }) => {
  await auth.use('api').authenticate()
  console.log(auth.use('api').user!)
})
```
So, how can I decode a token at any point throughout the code?

**Comment by aikrom** [comment]
It is not entirely clear to me which case you want to implement. Do you want to get information about the authorized user somewhere outside the route or controller?

**Reply by asta-nguyen** [reply]
yes, that is correctly I want

> It is not entirely clear to me which case you want to implement. Do you want to get information about the authorized user somewhere outside the route or controller?

Yes, that is exactly what I want.

**Reply by aikrom** [reply]
You can use [ALS](https://docs.adonisjs.com/guides/async-local-storage).

**Reply by aikrom** [reply]
But for most cases, it's enough to take information about the user in the controller and pass it through arguments to services or other functions that you want to use.

```ts
import { HttpContextContract } from '@ioc:Adonis/Core/HttpContext'

export default class PostsController {
  public async index({ auth }: HttpContextContract) {
    // Authenticate user or use route auth middleware to authenticate
    await auth.use('api').authenticate()

    // Do with user what you want in controller
    auth.user!

    // Or use in some service
    const someResult = SomeService.doSomething(auth.user!) 
  }
}
```

**Reply by asta-nguyen** [reply]
Thanks @aikrom
Another question: how do I decode a token sent by the client to the server adonisjs?

**Reply by aikrom** [reply]
I showed how to do it above.

```ts
await auth.use('api').authenticate()
```

That how you can decode token and authenticate user. 

Or it didn't work? Or you want only decode token and see token data without user authenticate?

**Reply by asta-nguyen** [reply]
I mean, if a client makes a request with a token in the body, how can we decode this token? I've tried to find a solution but am still stuck.

**Reply by aikrom** [reply]
You should send token in request headers.

Example with axios:

```ts
axios.get('/some/thing', {
    headers: {
       'Authorization': `Bearer ${token}` 
    }
})
```

**Reply by matvp91** [reply]
@aikrom there's cases where you're not relying on http, thus the HttpContextContract is not available (none of the controller logic is). Authentication with ws / socket.io for example. OP's question to decode a token beyond what's provided as part of the controller logic seems reasonable to me.

**Reply by Kuzzy** [reply]
I've did it by hands for socket.io, and I'm using Redis storage for tokens:

```typescript
// start/socket.ts
Ws.io
  .of('/myChannel')
  .on('connection', async socket => {
    const { auth, headers, query } = socket.handshake

    const token = auth.token.replace('Bearer ', '')
    const parts = token.split('.')
    const tokenId = base64.urlDecode(parts[0], undefined, true)
    const tokenStr = await Redis.get(`api:${tokenId}`)

    if (!tokenStr)
      throw new Exceptions.ForbiddenException(constants.errorMessages.forbiddenException)

    const tokenObj = JSON.parse(tokenStr)

    const userId = tokenObj.user_id

    // Now you can grab user from DB and do what you need
    let user = await User.findById(userId)

   ....

   // Additionally you can use Hash provider to verify token hash, something like that:
  const isValid = await Hash.verify(tokenObj.token, token)
})


// Client
import { Manager } from 'socket.io-client'

const manager = new Manager(wsUrl, {
        timeout: 25000,
        reconnectionDelay: 10000,
      })

const socket = manager.socket('/myChannel', {
        auth: {
          token, // Grab it from LS or from other place where it stored in client
        },
      })
```

Here is a source: https://github.com/adonisjs/auth/blob/a48cee194596d0096de38fba86f5b39f5a759861/src/Guards/Oat/index.ts#L196

And yes, it would be nice to share HttpContext somehow for such places to avoid extra coding and code duplication from sources.

**Reply by Rai-Satyam** [reply]
The above solution was very useful however I faced some issue with this:

While verifying the token in this following line

`const isValid = await Hash.verify(tokenObj.token, token)`

I got this error

`TypeError: The hash must be a valid phc string`

So I created my own function to compare the hash. Hope this will help someone stuck like me. :)

```
import crypto from 'crypto'

function compareToken(plainToken: string, hash: string) {
  const tokenHash = crypto.createHash('sha256').update(plainToken).digest('hex')

  return tokenHash === hash
}

const isValid = compareToken(token, tokenObj.token)
```

**Reply by aikrom** [reply]
https://github.com/adonisjs/core/issues/4146
